# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
def add_two(x):
    return x + 2

print(add_two)

<function add_two at 0x000002BBA60683A0>


In [2]:
#lambda param: one-line function

lambda x: x + 2

<function __main__.<lambda>(x)>

In [4]:
print(lambda x: x+2)

<function <lambda> at 0x000002BBA60685E0>


In [6]:
#IIFE - Immediately Invoked Function Expression
#Execute a lambda function immediately after defining the function
#Syntax(lambda param: expression)(arg)

(lambda x: x+2)(15)

17

#### Saving to a Variable

named_lambda = lambda z: z **2

In [8]:
named_lambda = lambda z: z ** 2

In [9]:
named_lambda(10)

100

#### Multiple Inputs

In [10]:
# Syntax: lambda param1, param2, etc: expression

def multiply(x, y, z):
    return x * y * z

multiply(10, 3, 2)

60

In [11]:
#AS an IIFE = (lambda param1, param2, etc: expression)(arg1, arg2, etc)

(lambda x,y,z:x*y*z)(10,3,2)

60

In [16]:
(lambda **kwargs: kwargs)(test=123, name='Aaron')

{'test': 123, 'name': 'Aaron'}

In [17]:
(lambda *args: args)('a','b','c')

('a', 'b', 'c')

In [ ]:
#Passing a Function into a Function

In [19]:
def apply_to_each(f, a_list):
    """
    Print the return value of function f applied to every element in the list
    f-> function to execute
    a_list -> List of elements to execute function f on
    """
    for element in a_list:
        f_return = f(element)
        print(f_return)

In [20]:
def get_first_letter(string):
    return string[0]

In [21]:
list_of_seasons = ['Winter','Spring','Summer','Fall']

In [22]:
apply_to_each(get_first_letter, list_of_seasons)

W
S
S
F


In [ ]:
def reverse_string(string):
    return string[::-1]

apply_to_each(reverse_string,  list_of_seasons)


#### Passing a Lambda into a Function

In [23]:
#call the apply_to_each function, pass in a lambda function as f argument

apply_to_each(lambda string: string[-1], list_of_seasons)

r
g
r
l


In [24]:
def square_num(x):
    return x**2

apply_to_each(square_num, [1,2,3,4,5])

1
4
9
16
25


In [27]:
def my_outer_function(name):
    print(f"Hello {name}")
    
    #define a locally scoped function
    def my_inner_function():
        print('This is the inner function')
        
    my_inner_function()
        
print(my_outer_function('eren'))

Hello eren
This is the inner function
None


In [28]:
my_outer_function_return()

NameError: name 'my_outer_function_return' is not defined

In [29]:
def create_multiplier(num):
    def multiply(x):
        return x * num
    return multiply

five_times = create_multiplier(5)
print(five_times)

<function create_multiplier.<locals>.multiply at 0x000002BBA61468B0>


In [30]:
five_times(10)

50

In [31]:
def new_multiplier(num):
    return lambda x: x * num

four = new_multiplier(4)
print(four)

<function new_multiplier.<locals>.<lambda> at 0x000002BBA6146940>


In [ ]:
four(10)

#### Returning a Lambda from a Function

#### If Statements within Lambdas

In [32]:
def double_pos_square_negs(number):
    if number >= 0:
        return number * 2
    else:
        return number ** 2
    
print(double_pos_square_negs(10))
print(double_pos_square_negs(20))
print(double_pos_square_negs(-10))
print(double_pos_square_negs(-20))

20
40
100
400


In [33]:
def double_pos_square_negs(number):
    return number *2 if number >= 0 else number ** 2
    
print(double_pos_square_negs(10))
print(double_pos_square_negs(20))
print(double_pos_square_negs(-10))
print(double_pos_square_negs(-20))

20
40
100
400


In [ ]:
(lambda num: num * 2 if num >= 0 else num ** 2)(10)

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [34]:
f = lambda x: x**3

print(f)

<function <lambda> at 0x000002BBA60688B0>


In [46]:
def my_generator(num):
    i = 0
    while i < num:
        yield i
        i += 1
x = my_generator(10)
print(x)

<generator object my_generator at 0x000002BBA7FB2F90>


In [47]:
for yielded_value in x:
    print(yielded_value)
    

0
1
2
3
4
5
6
7
8
9


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [49]:
def squared_minus_two(num):
    return num ** 2 - 2

numbers = [3, 5, 7, 2, 4, 3, 7, 6]

new_nums = []

for n in numbers:
    #apply function to element
    new_num = squared_minus_two(n)
    new_nums.append(new_num)
    
print(new_nums)

[7, 23, 47, 2, 14, 7, 47, 34]


In [51]:
list(map(squared_minus_two, numbers))

[7, 23, 47, 2, 14, 7, 47, 34]

#### Using Lambda's with Map

In [54]:
nums_again = [3, 7, 4, 9, 2, 6, 4, 3, 8]

list(map(lambda t: t**2 - 2, nums_again))

[7, 47, 14, 79, 2, 34, 14, 7, 62]

In [55]:
x_vals = [1,2,3,4]
y_vals = [10,20,30,40,50]

list(map(lambda a,b: a*b, x_vals, y_vals))

[10, 40, 90, 160]

In [61]:
my_string = 'hello world'


unicode = str(map(lambda letter: ord(letter), my_string))
print(unicode)

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [52]:
list(map(lambda x: x * 2 - 1, [1,2,3,4,5,6,7,8]))

[1, 3, 5, 7, 9, 11, 13, 15]

In [70]:
#Use the map function to double each number if it is even and halve it if it is odd

my_numbers = [1,2,3,4,5,6,7,8,9,10]

test = list(map(lambda x : x * 2 if x % 2 == 0 else x * 0.5, my_numbers))
print(test)

[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


In [ ]:
test = list(map(lambda x : x * 2 if x % 2 == 0 else x * 0.5, my_numbers))
print(test)

#Ask brian if you can convert map to dictionary


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [73]:
names = ['Bob', 'Billy','Alex','Barbara','Amy','Andy','Cahterine','Cameron']

def starts_with_a(name):
    return name[0].upper() == 'A'

a_names_map = list(filter(starts_with_a,names))
print(a_names_map)

['Alex', 'Amy', 'Andy']


In [74]:
def remainder(num):
    return num % 3

my_nums = [x for x in range (1,11)]
print(my_nums)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [75]:
list(map(remainder, my_nums))

[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]

#### Using Lambda's with Filter()

In [76]:
list(filter(lambda name: True if name[0] == 'B' else False, names))

['Bob', 'Billy', 'Barbara']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [94]:
from random import randint
import statistics

random_nums = [randint(1,100) for _ in range(20)]
print(statistics.mean(random_nums))

list(filter(lambda num: True if num > statistics.mean(random_nums) else False, random_nums))


44.6


[53, 96, 66, 59, 69, 58, 49, 72, 87, 97]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [95]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [96]:
def add_nums(x,y):
    print('='*20)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns {x + y}")
    return x + y
result_add = reduce(add_nums,[1,2,3,4,5])
print(result_add)

x: 1
y: 2
returns 3
x: 3
y: 3
returns 6
x: 6
y: 4
returns 10
x: 10
y: 5
returns 15
15


In [97]:
total = 0
for num in [1,2,3,4,5]:
    total += num
print(total)

15


#### Using Lambda's with Reduce()

In [ ]:
reduce(lambda x, y: x + y, [1, 2, 3, 4, 5])

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [99]:
reduce(lambda x, y: x * y, [1,2,3,4,5,6])

720

In [4]:
prod_total = 1
for num in [1,2,3,4,5]:
    prod_total *= num
    
print(prod_total)

120


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [10]:
import time

def capitalize(name):
    return name[0].upper() + name[1:].lower()

capitalize('eRen')

def full_name(first, last):
    #take in a first name, capitalize the first name, take in a last name, capilitaze the last name
    #returns concatenated string of first + space + last
    return capitalize(first) + " " + capitalize(last)

def greet_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")

greet_person('eren','enou','wednesday')


Hello Eren Enou, it sure is beautiful on this Wednesday


In [11]:
def func_one():
    print('Function One Started')
    time.sleep(5)
    print("Function One Ended")
    
def func_two():
    print("Function Two Started")
    func_one()
    print("Functions Two Ended")
    
func_two()

Function Two Started
Function One Started
Function One Ended
Functions Two Ended




#### Implementing a Base Case

In [15]:
def add_nums(num):
    #set up a base case
    if num <= 1:
        print(f"add_nums({num}) = {num}")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [ ]:
def factorial(num):
    if num <= 1:
        print(f"factorial({num}) = 1")
        return 1
    else:
        print(f"factorial({num}) = {num} * factorial({num - 1})")
        return num * factorial(num - 1)
    


In [2]:
cache = {}

def recursive_example(number):
    if number in cache:
        return cache[number]
    elif number <= 2:
        return number
    else:
        answer = recursive_example(number - 1) * recursive_example(number - 2)
        cache[number] = answer
        return answer
    
recursive_example(3)

2

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [3]:
def rec_sub(x):
    if x == 0:
        return rec_sub(x-1)

print(rec_sub(10))

None


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [4]:
# bad, never create infinite loops

#### Infinite Generator

In [6]:
lengths = map(len,['Brian','Stanton','Coding Temple'])

print(lengths)

In [7]:
for item in lengths:
    print(item)

5
7
13


In [8]:
for item in lengths:
    print(item)

In [9]:
help(iter)

Help on built-in function iter in module builtins:

iter(...)
    iter(iterable) -> iterator
    iter(callable, sentinel) -> iterator
    
    Get an iterator from an object.  In the first form, the argument must
    supply its own iterator, or be a sequence.
    In the second form, the callable is called until it returns the sentinel.



In [11]:
a_new_list = ['hi','hello','how are you','goodbye']

list_iter = iter(a_new_list)
print(list_iter)

In [12]:
next(list_iter)

'hi'

In [14]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        return start
        start += stop
        
my_range_iter(0,10)

0

In [15]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        yield start
        start += stop
        
my_range_iter(0,10)

<generator object my_range_iter at 0x000002B85A89DC10>

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [20]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

places = list(filter(lambda x: x.strip() !='', places))

print(places)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [22]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key = lambda last_name: last_name.lower().split()[-1])

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


In [16]:
help(list.sort)

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [33]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]


f_places = list(map(lambda x: (x[0], (9/5)*x[1]+32), places))
print(f_places)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [38]:
def fib(n):
    if n <= 1:
        return 1
    else:
        return fib(n-1) + fib(n-2)

In [43]:
print(fib(10))

89


In [17]:
#Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

SyntaxError: invalid syntax (1237481684.py, line 1)

In [45]:
def my_gen(number):
    while number > 0:
        yield number ** 2
        number -= 1

for num in my_gen(10):
    print(num)
    
#100
#81
#64
#49
#36
#25
#16
#9
#4
#1


100
81
64
49
36
25
16
9
4
1
